### PR TITLE
 feat: 0x1B - SHL opcode

### PR DESCRIPTION
**Pull request by trbutler4**
implemented the shift left operator
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #15 

## What is the new behavior?
shift left operator implemented using utils::math wrapping shl 

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: crates/evm/src/tests/test_instructions/test_comparison_operations.cairo [hunk 1]
```diff
@@ -213,6 +213,43 @@ fn test_exec_gt_true() {
...
+    // Given 
+    let mut ctx = setup_execution_context();
+    ctx.stack.push(0xff00000000000000000000000000000000000000000000000000000000000000).unwrap();
+    ctx.stack.push(4_u256).unwrap();
+
+    // When 
+    ctx.exec_shl();
+
+    // Then
+    assert(ctx.stack.len() == 1, 'stack should have one element');
```
Comment (enitrat) [inline crates/evm/src/tests/test_instructions/test_comparison_operations.cairo:228]: you can group both assertions in the same test case. 
For a single test case with common test data, we must verify two different assertions, so it doesn't really make sense to separate the assertions

**Review by enitrat — COMMENTED** [review]
Can you introduce a test case for `shift>255` as well?

File: crates/evm/src/tests/test_instructions/test_comparison_operations.cairo (hunk 1 above)
Comment (trbutler4) [reply crates/evm/src/tests/test_instructions/test_comparison_operations.cairo:228]: i misunderstood what you meant on the previous pr when you asked me to separate into two tests with explicit naming. Put both assertions back

Comment (trbutler4) [reply crates/evm/src/tests/test_instructions/test_comparison_operations.cairo:228]: also added test for shift > 255

Comment (trbutler4) [reply crates/evm/src/tests/test_instructions/test_comparison_operations.cairo:228]: also added test for shift > 255

File: crates/evm/src/tests/test_instructions/test_comparison_operations.cairo [hunk 2]
```diff
@@ -213,6 +213,50 @@ fn test_exec_gt_true() {
     assert(ctx.stack.peek().unwrap() == 1, 'stack top should be 1');
 }
 
+#[test]
+#[available_gas(20000000)]
+fn test_exec_shl_wrapping() {
```
Comment (enitrat) [inline crates/evm/src/tests/test_instructions/test_comparison_operations.cairo:218]:
```suggestion
fn test_exec_shl() {
```

File: crates/evm/src/tests/test_instructions/test_comparison_operations.cairo [hunk 3]
```diff
@@ -213,6 +213,50 @@ fn test_exec_gt_true() {
...
+            .stack
+            .peek()
+            .unwrap() == 0xf000000000000000000000000000000000000000000000000000000000000000,
+        'stack top should be 0xf00000...'
+    );
+}
+
+#[test]
+#[available_gas(20000000)]
+fn test_exec_shl_shift_too_large() {
```
Comment (enitrat) [inline crates/evm/src/tests/test_instructions/test_comparison_operations.cairo:240]:
```suggestion
fn test_exec_shl_wrapping() {
```

**Review by enitrat — DISMISSED** [review]

**Review by enitrat — APPROVED** [review]

File: crates/evm/src/instructions/comparison_operations.cairo [hunk 4]
```diff
@@ -116,7 +116,17 @@ impl ComparisonAndBitwiseOperations of ComparisonAndBitwiseOperationsTrait {
     /// 0x1B - SHL
     /// # Specification: https://www.evm.codes/#1b?fork=shanghai
     fn exec_shl(ref self: ExecutionContext) -> Result<(), EVMError> {
-        Result::Ok(())
+        let popped = self.stack.pop_n(2)?;
+        let shift = *popped[0];
+        let val = *popped[1];
+
+        // if shift is bigger than 255 return 0
```
Comment (ClementWalter) [inline crates/evm/src/instructions/comparison_operations.cairo:123]: is this only to save gas? if so, could you add a comment

**Comment by ClementWalter** [comment]
just a questino regarding the early return

**Comment by trbutler4** [comment]
> just a questino regarding the early return

the reason for the early return was just to match the evm opcode specification: https://www.evm.codes/?fork=shanghai

where it says: value << shift: the shifted value. If shift is bigger than 255, returns 0.

should note that the shr opcode includes the same specification, but does not have an early return. Maybe needs to be updated.

**Pull request merged**